### DISCUSSION TITLE
useMemo in a server component?

### DISCUSSION THREAD

**Discussion by pomy4** [vercel/next.js#68071] [Help] [answered]
### Summary

Is there way to cache a value in a server component so that when it is rendered for the second time (on router.refresh()) it reuses the value from the first render? Please see the code below for a better explanation. It seems to me this isn't possible, but hopefully I'm wrong.

### Additional information

```js
### server component ------------------------------------------

import Client from './client';

async function fetchData() {
    console.log('expensive operation');
    return 'data';
}

export default async function Server() {
    // is there a way to avoid calling this function
    // if the component is not being rendered for the first time
    // something like this
    // const data = await useMemo(fetchData, []);
    const data = await fetchData();
    return <Client data={data}/>;
}

### client component ------------------------------------------

"use client";

import { useState } from "react";
import { useRouter } from "next/navigation";

export default function Client({ data: initialData }) {
  const [data, setData] = useState(initialData);
  const router = useRouter();
  return (
    <div>
      <div>{data}</div>;
      <button onClick={() => router.refresh()}>refresh</button>
    </div>
  );
}
```

### Example

https://codesandbox.io/p/devbox/spring-pine-sjlgpg?workspaceId=2ed7787a-376d-4fad-ae44-cbb614a91b66

**Comment by icyJoseph** [comment, marked as answer]
Look at: https://nextjs.org/docs/app/api-reference/functions/unstable_cache

**Reply by pomy4** [reply]
Thanks, this looks good.

**Comment by romhenri** [comment]
Hooks are to React Client Components, to server, i think you should use cache.

**Reply by icyJoseph** [reply]
One quirky thing with `useMemo`, `useId`, `useCallback`, is that, they can work in Server Components, although uselessly if you will, see this example:

- https://stackblitz.com/edit/nextjs-drwlhf?file=app%2Fpage.tsx

As soon as you try to work with `useState`, or other type of interactivity, the page breaks, but `useMemo` and `useCallback` are sort of a `noop` (unnecessary but workable), at least on these versions. 

`useId` is the cool useful example, because, it does generate an id, that you can then put in an element, and it works in server components too (at least in the React versions I've used), and so far I haven't experienced any issues with mismatched ids and stuff.
